File: backend/app/models/inventory.py
import uuid
from datetime import datetime, timezone
from decimal import Decimal
from typing import Optional, TYPE_CHECKING
from enum import Enum

from sqlalchemy import Column, DateTime, Numeric, String, UniqueConstraint, CheckConstraint
from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from app.models.workspace import Workspace
    from app.models.product import Product

class Warehouse(SQLModel, table=True):
    __tablename__ = "warehouses"
    __table_args__ = (
        UniqueConstraint("workspace_id", "code", name="uq_workspace_warehouse_code"),
    )
    
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    workspace_id: uuid.UUID = Field(foreign_key="workspaces.id", nullable=False, index=True)
    
    code: str = Field(max_length=50, index=True)
    name: str = Field(max_length=255)
    location: Optional[str] = Field(default=None, max_length=255)
    is_active: bool = Field(default=True)
    
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        sa_column=Column(DateTime(timezone=True), nullable=False)
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        sa_column=Column(DateTime(timezone=True), nullable=False)
    )

class WarehouseBin(SQLModel, table=True):
    __tablename__ = "warehouse_bins"
    __table_args__ = (
        UniqueConstraint("warehouse_id", "code", name="uq_warehouse_bin_code"),
    )
    
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    warehouse_id: uuid.UUID = Field(foreign_key="warehouses.id", nullable=False, index=True)
    
    code: str = Field(max_length=50, index=True)
    barcode: Optional[str] = Field(default=None, max_length=100)
    is_active: bool = Field(default=True)

class InventoryLevel(SQLModel, table=True):
    """
    Real-time snapshot of inventory at the Bin level.
    Rule 1.3: Available Stock = on_hand - reserved - damaged
    """
    __tablename__ = "inventory_levels"
    __table_args__ = (
        UniqueConstraint("product_id", "bin_id", name="uq_inventory_product_bin"),
        CheckConstraint("on_hand >= 0", name="chk_inventory_on_hand_positive"),
        CheckConstraint("reserved >= 0", name="chk_inventory_reserved_positive"),
        CheckConstraint("damaged >= 0", name="chk_inventory_damaged_positive"),
    )
    
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    workspace_id: uuid.UUID = Field(foreign_key="workspaces.id", nullable=False, index=True)
    product_id: uuid.UUID = Field(foreign_key="products.id", nullable=False, index=True)
    warehouse_id: uuid.UUID = Field(foreign_key="warehouses.id", nullable=False, index=True)
    bin_id: uuid.UUID = Field(foreign_key="warehouse_bins.id", nullable=False, index=True)
    
    on_hand: Decimal = Field(default=0, sa_column=Column(Numeric(12, 2), nullable=False))
    reserved: Decimal = Field(default=0, sa_column=Column(Numeric(12, 2), nullable=False))
    damaged: Decimal = Field(default=0, sa_column=Column(Numeric(12, 2), nullable=False))
    
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        sa_column=Column(DateTime(timezone=True), nullable=False)
    )

class TransactionType(str, Enum):
    INITIAL = "INITIAL"
    RECEIPT = "RECEIPT"
    ISSUE = "ISSUE"
    TRANSFER = "TRANSFER"
    ADJUSTMENT = "ADJUSTMENT"

class InventoryTransaction(SQLModel, table=True):
    """
    Rule 1.2: StockTransaction is an IMMUTABLE ledger. All changes flow through business events.
    """
    __tablename__ = "inventory_transactions"
    
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    workspace_id: uuid.UUID = Field(foreign_key="workspaces.id", nullable=False, index=True)
    product_id: uuid.UUID = Field(foreign_key="products.id", nullable=False, index=True)
    
    transaction_type: TransactionType = Field(nullable=False)
    
    # Positive for Receipt/In, Negative for Issue/Out
    quantity: Decimal = Field(sa_column=Column(Numeric(12, 2), nullable=False))
    
    source_bin_id: Optional[uuid.UUID] = Field(default=None, foreign_key="warehouse_bins.id")
    destination_bin_id: Optional[uuid.UUID] = Field(default=None, foreign_key="warehouse_bins.id")
    
    reference_type: Optional[str] = Field(default=None, max_length=50) # e.g. GRN, DO, INVOICE
    reference_id: Optional[uuid.UUID] = Field(default=None) # Foreign key to the exact document
    
    user_id: uuid.UUID = Field(foreign_key="users.id", nullable=False)
    
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        sa_column=Column(DateTime(timezone=True), nullable=False)
    )

File: frontend/src/api/inventory.ts
import { apiClient } from './client';
import type { SuccessResponse } from '../types/api';

export interface Warehouse {
  id: string;
  code: string;
  name: string;
  location: string | null;
  is_active: boolean;
  created_at: string;
}

export interface WarehouseBin {
  id: string;
  warehouse_id: string;
  code: string;
  barcode?: string | null;
  is_active: boolean;
}

export interface InventoryLevel {
  id: string;
  product_id: string;
  warehouse_id: string;
  bin_id: string;
  on_hand: number;
  reserved: number;
  damaged: number;
  available: number;
}

export const getWarehouses = async (): Promise<Warehouse[]> => {
  const response = await apiClient.get<SuccessResponse<Warehouse[]>>('/api/v1/inventory/warehouses');
  return response.data.data;
};

export const getWarehouseBins = async (warehouseId: string): Promise<WarehouseBin[]> => {
  const response = await apiClient.get<SuccessResponse<WarehouseBin[]>>(
    `/api/v1/inventory/warehouses/${warehouseId}/bins`,
  );
  return response.data.data;
};

export const createWarehouse = async (data: Partial<Warehouse>): Promise<Warehouse> => {
  const response = await apiClient.post<SuccessResponse<Warehouse>>('/api/v1/inventory/warehouses', data);
  return response.data.data;
};

export const getInventoryLevels = async (productId?: string): Promise<InventoryLevel[]> => {
  const url = productId ? '/api/v1/inventory/levels?product_id=' + productId : '/api/v1/inventory/levels';
  const response = await apiClient.get<SuccessResponse<InventoryLevel[]>>(url);
  return response.data.data;
};
